### PR TITLE
Added "harmony-lemma-formalization" as a new case study

### PR DESCRIPTION
**Pull request by GabrieleCecilia**
I am providing the Beluga code for the paper "A Beluga Formalization of the Harmony Lemma in the π-Calculus" by Gabriele Cecilia and Alberto Momigliano, so that it can be revised, stored as a case study and eventually updated.

### CODE REVIEW

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+rec fstep_impl_red: (g:ctx) [g ⊢ fstep P f_tau Q] → [g ⊢ P red Q] =
+/ total f (fstep_impl_red _ _ _ f) /
+fn f ⇒ case f of
+   | [g ⊢ fs_par1 F1] ⇒ let [g ⊢ R] = fstep_impl_red [g ⊢ F1] in
+                         [g ⊢ r_par R]
+   | [g ⊢ fs_par2 F2] ⇒ let [g ⊢ R] = fstep_impl_red [g ⊢ F2] in
+                         [g ⊢ r_str par_comm (r_par R) par_comm] 
+   | [g ⊢ fs_com1 F1 B1] ⇒
+     let [g ⊢ D1] = fs_out_rew [g ⊢ _] [g ⊢ F1] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:198]:
```suggestion
     let [g ⊢ D1] = fs_out_rew [g ⊢ F1] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ D2] = bs_in_rew [g ⊢ _] [g ⊢ B1] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:199]:
```suggestion
     let [g ⊢ D2] = bs_in_rew [g ⊢ B1] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ R] = fs_com1_impl_red [g ⊢ D1] [g ⊢ D2] in [g ⊢ R]
+   | [g ⊢ fs_com2 B1 F1] ⇒
+     let [g ⊢ D1] = bs_in_rew [g ⊢ _] [g ⊢ B1] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:202]:
```suggestion
     let [g ⊢ D1] = bs_in_rew [g ⊢ B1] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ D2] = fs_out_rew [g ⊢ _] [g ⊢ F1] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:203]:
```suggestion
     let [g ⊢ D2] = fs_out_rew [g ⊢ F1] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ R] = fs_com2_impl_red [g ⊢ D1] [g ⊢ D2] in [g ⊢ R]
+   | [g ⊢ fs_res \z.F[..,z]] ⇒
+     let [g,z:names ⊢ R[..,z]] = fstep_impl_red [g,z:names ⊢ F[..,z]] in 
+     [g ⊢ r_res \z.R[..,z]]
+   | [g ⊢ fs_close1 B1 B2] ⇒
+     let [g ⊢ D1] = bs_out_rew [g ⊢ _][g ⊢ B1] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:209]:
```suggestion
     let [g ⊢ D1] = bs_out_rew [g ⊢ B1] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ D2] = bs_in_rew [g ⊢ _] [g ⊢ B2] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:210]:
```suggestion
     let [g ⊢ D2] = bs_in_rew [g ⊢ B2] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ R] = fs_close1_impl_red [g ⊢ D1] [g ⊢ D2] in [g ⊢ R]
+   | [g ⊢ fs_close2 B1 B2] ⇒
+     let [g ⊢ D1] = bs_in_rew [g ⊢ _] [g ⊢ B1] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:213]:
```suggestion
     let [g ⊢ D1] = bs_in_rew [g ⊢ B1] in
```

File: case-studies/harmony-lemma-formalization/5_theorem1.bel [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+     let [g ⊢ D2] = bs_out_rew [g ⊢ _] [g ⊢ B2] in
```
Comment (MartyO256) [inline case-studies/harmony-lemma-formalization/5_theorem1.bel:214]:
```suggestion
     let [g ⊢ D2] = bs_out_rew [g ⊢ B2] in
```

**Review by MartyO256 — CHANGES_REQUESTED** [review]
Type-checking of the file `5_theorem1.bel` failed during continuous integration. This is due to an extra argument `[g ⊢ _]` given to the functions `fs_out_rew`, `bs_in_rew` and `bs_out_rew` in `fstep_impl_red`.

**Comment by MartyO256** [comment]
Thank you for your contribution! Incorporating your case study will help us improve the implementation of Beluga.

Please let me know if the changes I requested above are acceptable.

**Comment by GabrieleCecilia** [comment]
Thank you for your work: you are right about the mistakes. This is because I actually uploaded an update of the functions `bs_in_rew`, `fs_out_rew` and `bs_out_rew` (in which I removed an unnecessary explicit argument and performed a more efficient induction) without remembering to update the function `fstep_impl_red` as well. Correctness of my code is something I should have checked beforehand: I apologize for being superficial.

In conclusion: you can apply the changes you suggested. After applying them, the code is correct.
